### PR TITLE
Proof of concept: Markdown metadata

### PR DESCRIPTION
**Pull request by mootari**
**NOTE: Do not merge, only provided for review and discussion.**

Goals:
- Provide a single source of truth for managing paths, slugs and redirects.
- Use (inherited) slugs to decouple URLs from file names while inheriting the content hierarchies.
- Stop the redirect mess by managing redirects primarily in the files they are supposed to point to. `public/html/_redirects` will be generated at build time from markdown metadata.
- Enhance editing capabilities by incorporating https://github.com/markdown-it/markdown-it-container.
- Provide entry points to all directories by adding `index.md` files in each.
- Encourage content polishing (e.g. title, abstract) by allowing contributors to edit a *single* file (instead of additional configs all over the place).

Progress:
- Migrated metadata from statics.json and _redirects into Markdown files (via script, reproducable).

### CODE REVIEW

**Comment by mootari** [comment]
@deltakosh This is part of an effort to simplify the build pipeline and overall documentation structure. Side note: I'm currently in the process of migrating to assemble and handlebars, which is basically unavoidable if I want to have any chance of properly integrating TypeDoc.

I'd love to hear your feedback. :)

**Comment by mootari** [comment]
... Btw, I'm also planning to add "Edit this page" links so that anyone with a github account can click on it to quickly create Edit pull requests.

**Comment by deltakosh** [comment]
thanks a lot! I'll have a deep look tomorrow
For those interested in chiming in: https://deploy-preview-737--babylonjsdoc.netlify.com/

File: content/babylon101/index.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+title: All
+abstract: The How To of 101.
+slug: babylon-101
```
Comment (mootari) [inline content/babylon101/index.md:4]: Duplicate slug.

File: content/How_To/GUI/index.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+title: GUI
+abstract: Babylon.GUI
+slug: gui
```
Comment (mootari) [inline content/How_To/GUI/index.md:4]: Duplicate slug.

**Review by mootari — COMMENTED** [review]
The future tooling will definitely have to detect duplicate slugs (although that shouldn't be hard to do). Luckily these problems will be only localized as slugs inherit /add to their parent path.

**Comment by mootari** [comment]
I've also started to decouple the menu structure. Current declaration looks like this:
```javascript
const menuMain = [
  { label: 'Babylon.JS',    path: 'http://www.babylonjs.com/', icon: 'play' },
  { label: 'Documentation', path: '/',           icon: 'home' },
  { label: "What's new",    path: '/whats-new',  icon: 'file-text-o' },
  { label: 'Babylon 101',   path: '/babylon101', icon: 'book' },
  { label: 'How To ...',    path: '/How_To',     icon: 'cogs' },
  { label: 'Features',      path: '/features',   icon: 'book' },
  { label: 'Resources',     path: '/resources',  icon: 'rocket' },
  { label: 'Extensions',    path: '/extensions', icon: 'wrench' },
  { label: 'Samples',       path: '/samples',    icon: 'book' },
  { label: 'Classes',       path: '/classes',    icon: 'files-o' },
  { label: 'Playground',    path: '/playground', icon: 'play' },
];

const menuFooter = [
  { label: 'Forum',               path: 'http://www.html5gamedevs.com/forum/16-babylonjs', icon: 'html5' },
  { label: 'Github',              path: 'https://github.com/BabylonJS/Babylon.js', icon: 'github' },
  { label: 'Contribute',          path: 'https://github.com/BabylonJS/Documentation', icon: 'code-fork' },
  { label: 'Deployed by netlify', path: 'https://www.netlify.com/', icon: 'heart' },
];
```
I don't expect these paths to be validated, but I doubt that someone will accidentally screw with them without anybody noticing.

**Comment by deltakosh** [comment]
I'm about to merge. More stuff to add?

**Comment by mootari** [comment]
Woah, hold on! This one really shouldn't be merged. It will completely break the current docs (e.g. the links do not yet match the actual structure).

It's just meant as a preview of how metadata could be managed in the future, and what links could look like. Once the tooling is complete and we've resolved all the glitches and gotchas we can run a new update for master.

Some of the stuff that's still missing:
- Actual content for the index files
- Metadata defaults
- Removal of PG_TITLE and ID_PAGE
- Generate files using the new slugs

**Comment by deltakosh** [comment]
Was kidding :) I should have added a smiley :D

**Comment by RaananW** [comment]
So, I am merging.

**Comment by mootari** [comment]
"redirects" should be changed to "redirect_from" to match [Github's redirect feature](https://help.github.com/articles/redirects-on-github-pages/). The list style already matches the [Jekyll syntax](https://github.com/jekyll/jekyll-redirect-from). It's also less ambiguous.

**Comment by mootari** [comment]
We also need a good way to deal with moved/renamed files and directories, ideally via atomic tasks:
- `redirect:add-current`: Adds each file's current slug/url to its redirects.
- `redirect:remove-current`: Removes redirects that match a file's current slug/url.
- `update-links`: Compiles and resolves all redirects, then updates local links in-place in markdown files.
- `rename:prepare`: Runs `redirect:add-current`.
- `rename:finish`: Runs `redirect:remove-current` and `update-links`

We need to ensure that running `redirect:remove-current` directly after `redirect:add-current` leaves the files effectively unchanged (I've seen some cases where people seemingly committed stuff without closer inspection). One exception would be formatting/order changes after someone edited frontmatter manually in an earlier commit (ideally we'd have a sanitize task for that).

The resulting workflow would be:
1. Run `rename:prepare`.
2. Move files around, rename files or slugs, go wild.
3. Run `rename:finish`.
5. Add and commit.

**Comment by mootari** [comment]
Required tests (after build):
- No broken redirects
- No broken internal links (not considering redirects, optionally excluding API pages)
- No broken/missing images
- Every directory that contains .html files has an index.html

Additionally a sitemap.xml and (optionally) a public redirects.json should be generated. Both can be consumed in BabylonJS tests to check for dead links in annotations.

**Comment by deltakosh** [comment]
Site map is a must as it will help for SEO

**Comment by mootari** [comment]
> Site map is a must as it will help for SEO

I agree. I've also added dynamic page titles and included abstracts (where available) as meta descriptions. Additionally I'm considering adding abstracts as title attributes to menu links, but I also don't want to bloat the generated markup too much.

**Comment by deltakosh** [comment]
We are not good SEO wise so everything that can help is welcome

**Comment by mootari** [comment]
I've forcefully disabled browser autocompletion for the search field. The suggestions were a mix of search terms from various sites that all used the same input name, resulting in terrible UX (because there are no results for those terms).

I've documented this decision extensively in the template though.

**Edit:** ... or rather, I've *tried*, but Chrome is having none of it. I've created #750 to investigate other options.

**Comment by deltakosh** [comment]
We should not turn it off anyway but instead make sure it works with our scope only :)

**Comment by mootari** [comment]
A small sneak peek of the improved playground embeds: https://imgur.com/mCd9qJI

**Comment by deltakosh** [comment]
Huge..love it!

**Comment by mootari** [comment]
Regarding "What's new": Is there a strong desire to show changelogs for yet unreleased versions (e.g. 3.2.0 / preview release)? Otherwise I could simply fetch the changelogs via Github's REST API for releases, which would make things very simple.

**Comment by deltakosh** [comment]
This is even better :)

**Comment by deltakosh** [comment]
BTW do you have an ETA of a first drop you an do (mostly regarding classes Doc generation?)
I want to get rid of the current tool as it is too much buggy 
If you cannot it is not a big deal and we will certainly unblock some time to work on it asap

**Comment by mootari** [comment]
I'd have to be conservative and say another two weeks. I've basically rewritten everything, and need to reintegrate TypeDoc using its JSON output so that it works nicely with the rest of the Assemble setup.

I've prioritized the changelog because it allows me to test how I need to deal with remote resources. After that I'll move on to TypeDoc, and after *that* I need to tie everything together.

It's gonna be good (I think), but it's a ****load more work than I had anticipated.

**Comment by deltakosh** [comment]
Can we help on something?

**Comment by mootari** [comment]
Not yet I'm afraid. Give it another week and I'll probably be able to stomp out a preview branch that's workable.

**Comment by deltakosh** [comment]
Sounds good
Thanks a lot

**Comment by mootari** [comment]
No problem, I'm learning a ton of stuff along the way. 😄

**Comment by deltakosh** [comment]
And you are helping a great deal!

**Comment by mootari** [comment]
Btw, how come you're publishing preleases on npm, but not on github?

**Comment by deltakosh** [comment]
We are publishing on both. What do you mean?

**Comment by mootari** [comment]
> We are publishing on both. What do you mean?

[3.2.0-alpha4 on npm](https://www.npmjs.com/package/babylonjs) vs [3.1.0 on Github](https://github.com/BabylonJS/Babylon.js/releases). On Github v3.1.0 is also the most recent tag.

**Comment by deltakosh** [comment]
We are not tagging every version (we should cc @raananw)

**Comment by RaananW** [comment]
@deltakosh - I can tag when commiting the npm updates.

**Comment by deltakosh** [comment]
That would be cool (and update the engine version if I forget:))

**Comment by RaananW** [comment]
will do that starting next npm push. The publisher can update the engine.ts file as well to keep it consistent.

**Comment by mootari** [comment]
Btw, I just finished the changelogs and noticed some unexpected positive side effects 🎉
bildschirmfoto 2018-01-18 um 15 56 59

**Comment by deltakosh** [comment]
Yeah:)

**Comment by mootari** [comment]
The data passed to the template also contains the [complete release info](https://api.github.com/repos/BabylonJS/Babylon.JS/releases/tags/v3.1.0) for each release, so we can directly link to Github or even provide download links etc

**Comment by deltakosh** [comment]
Just wondering if we should not use directly the html output of typedoc ?
THe html is really beautiful

**Comment by mootari** [comment]
> Just wondering if we should not use directly the html output of typedoc ?
THe html is really beautiful

I can certainly try to reuse the styles and templates. Beyond that using the TypeDoc generation process directly creates several integration issues that are hard to work around.

Another thing to keep in mind: The current API docs target consumers of the JS build by excluding TypeScript specific features like interfaces. I had those ripped out already when I used typedoc directly (before I started rewriting literally everything).

My suggestion would be that I keep as much as possible of the original TypeDoc output for the first preview. We can then compare them to the current docs and decide what needs to be dropped.

I've already implemented remote d.ts fetching and conversion to JSON and also wrote a simple file cache which speeds up both operations significantly on subsequent builds. Next I'll take a closer look at how I can make the JSON work with the templates - which might be tricky since the themes normally consume Reflection instances that are only created during conversion and cannot be cached/persisted.

Btw, the currently defined API sources are v3.1.0, v.3.0.7 and nightly.

**Comment by deltakosh** [comment]
Yeah let's keep the most of Typedoc output as it is pretty beautiful (and far far far better than what we have now)
For API sources please only keep stable and preview

**Comment by deltakosh** [comment]
Hello hello:)
any news on your battle front ?

**Comment by deltakosh** [comment]
Hello? Do you need some help?
I would like to enable JSdoc for 3.2 release so let me know if you need help

**Comment by mootari** [comment]
Hi, sorry for the late reply. Had to pause working on this because work/life got in the way.

What exactly do you mean by "enable JSdoc"? As far as I'm aware annotations wouldn't survive the transpilation of TS to JS?

My current status is that I had migrated the theme to [Bulma](https://bulma.io/) to have a better basis for integration of the typedoc theme. I was also working on making the typedoc default theme play nice with typedoc JSON output (instead of reflection instances).
A semi-functional pre-bulma version ist committed to [my fork]( https://github.com/mootari/Documentation/tree/rewrite-generator-assemble), but I haven't pushed the more recent changes.

All in all there's still a lot of work to be done.

**Comment by deltakosh** [comment]
I meant Typedoc sorry
I would like to really update the classes doc

**Comment by deltakosh** [comment]
Hello any news??

**Comment by deltakosh** [comment]
Closing for now as things evolved recently to rely on typedoc
Please open a new PR for others great updates you have here

**Pull request closed without merging**